File: CS/Task26/VistaMyDateEditCalendar.cs
using System;
using System.Collections.Generic;
using System.Windows.Forms;
using DevExpress.XtraEditors.Controls;
using DevExpress.XtraEditors.Calendar;
using DevExpress.Utils.Menu;
using DevExpress.Utils;

namespace Task26
{
    public class VistaMyDateEditCalendar : VistaDateEditCalendar
    {
        ContextMenu fNoteMenu;

        public VistaMyDateEditCalendar(RepositoryItemMyDateEdit item, object editDate)
            : base(item, editDate)
        {
            fNoteMenu = CreateNoteMenu();
        }

        public new RepositoryItemMyDateEdit Properties
        {
            get { return base.Properties as RepositoryItemMyDateEdit; }
        }

        protected virtual ToolTipController TipController
        {
            get { return ToolTipController.DefaultController; }
        }

        protected override DevExpress.XtraEditors.Drawing.DateEditPainter CreatePainter()
        {
            return new VistaMyDateEditCalendarPainter(this);
        }

        protected override void OnMouseUp(System.Windows.Forms.MouseEventArgs e)
        {
            if (e.Button == MouseButtons.Right)
            {
                CalendarHitInfo HI = GetHitInfo(e);
                if (HI.InfoType == CalendarHitInfoType.Item)
                {
                    this.DateTime = HI.HitDate;
                    bool NotEmpty = Properties.HaveNote(HI.HitDate);
                    fNoteMenu.MenuItems[1].Enabled = NotEmpty;
                    fNoteMenu.MenuItems[2].Enabled = NotEmpty;
                    fNoteMenu.Show(this, e.Location);
                }
            }
            else
                base.OnMouseUp(e);
        }

        protected override void OnHotObjectChanged(CalendarHitInfo prevInfo, CalendarHitInfo currInfo)
        {
            base.OnHotObjectChanged(prevInfo, currInfo);
            if (currInfo.InfoType == CalendarHitInfoType.Item)
            {
                List<Note> Notes = Properties.GetDateNotes(currInfo.HitDate);
                if (Notes.Count > 0)
                {
                    string NotesList = "";
                    foreach (Note N in Notes)
                    {
                        NotesList += N.Date.ToShortTimeString() + ": " + N.Name + '\n';
                    }
                    TipController.ShowHint(NotesList, currInfo.HitDate.ToShortDateString() + ": " + Notes.Count + " Notes.", PointToScreen(currInfo.Pt));
                    return;
                }
            }
            TipController.HideHint();
        }
        
        protected virtual ContextMenu CreateNoteMenu()
        {
            ContextMenu CM = new ContextMenu();
            CM.MenuItems.Add(new MenuItem("Add Note...", new EventHandler(OnMenuAddNote_Click)));
           
            CM.MenuItems.Add(new MenuItem("Edit...", new EventHandler(OnMenuEditNote_Click)));
            CM.MenuItems.Add(new MenuItem("Delete...", new EventHandler(OnMenuDeleteNote_Click)));

            return CM;
        }

        protected virtual void OnMenuAddNote_Click(object sender, EventArgs e)
        {
            Properties.AddNote(DateTime);
        }

        protected virtual void OnMenuEditNote_Click(object sender, EventArgs e)
        {
            Properties.EditNotes(DateTime);
        }

        protected virtual void OnMenuDeleteNote_Click(object sender, EventArgs e)
        {
            Properties.DeleteNotes(DateTime);
        }
    }
}

File: VB/Task26/VistaMyDateEditCalendar.vb
Imports Microsoft.VisualBasic
Imports System
Imports System.Collections.Generic
Imports System.Windows.Forms
Imports DevExpress.XtraEditors.Controls
Imports DevExpress.XtraEditors.Calendar
Imports DevExpress.Utils.Menu
Imports DevExpress.Utils

Namespace Task26
	Public Class VistaMyDateEditCalendar
		Inherits VistaDateEditCalendar
		Private fNoteMenu As ContextMenu

		Public Sub New(ByVal item As RepositoryItemMyDateEdit, ByVal editDate As Object)
			MyBase.New(item, editDate)
			fNoteMenu = CreateNoteMenu()
		End Sub

		Public Shadows ReadOnly Property Properties() As RepositoryItemMyDateEdit
			Get
				Return TryCast(MyBase.Properties, RepositoryItemMyDateEdit)
			End Get
		End Property

		Protected Overridable ReadOnly Property TipController() As ToolTipController
			Get
				Return ToolTipController.DefaultController
			End Get
		End Property

		Protected Overrides Function CreatePainter() As DevExpress.XtraEditors.Drawing.DateEditPainter
			Return New VistaMyDateEditCalendarPainter(Me)
		End Function

		Protected Overrides Sub OnMouseUp(ByVal e As System.Windows.Forms.MouseEventArgs)
			If e.Button = MouseButtons.Right Then
				Dim HI As CalendarHitInfo = GetHitInfo(e)
				If HI.InfoType = CalendarHitInfoType.Item Then
					Me.DateTime = HI.HitDate
					Dim NotEmpty As Boolean = Properties.HaveNote(HI.HitDate)
					fNoteMenu.MenuItems(1).Enabled = NotEmpty
					fNoteMenu.MenuItems(2).Enabled = NotEmpty
					fNoteMenu.Show(Me, e.Location)
				End If
			Else
				MyBase.OnMouseUp(e)
			End If
		End Sub

		Protected Overrides Sub OnHotObjectChanged(ByVal prevInfo As CalendarHitInfo, ByVal currInfo As CalendarHitInfo)
			MyBase.OnHotObjectChanged(prevInfo, currInfo)
			If currInfo.InfoType = CalendarHitInfoType.Item Then
				Dim Notes As List(Of Note) = Properties.GetDateNotes(currInfo.HitDate)
				If Notes.Count > 0 Then
					Dim NotesList As String = ""
					For Each N As Note In Notes
						NotesList &= N.Date.ToShortTimeString() & ": " & N.Name + ControlChars.Lf
					Next N
					TipController.ShowHint(NotesList, currInfo.HitDate.ToShortDateString() & ": " & Notes.Count & " Notes.", PointToScreen(currInfo.Pt))
					Return
				End If
			End If
			TipController.HideHint()
		End Sub

		Protected Overridable Function CreateNoteMenu() As ContextMenu
			Dim CM As New ContextMenu()
			CM.MenuItems.Add(New MenuItem("Add Note...", New EventHandler(AddressOf OnMenuAddNote_Click)))

			CM.MenuItems.Add(New MenuItem("Edit...", New EventHandler(AddressOf OnMenuEditNote_Click)))
			CM.MenuItems.Add(New MenuItem("Delete...", New EventHandler(AddressOf OnMenuDeleteNote_Click)))

			Return CM
		End Function

		Protected Overridable Sub OnMenuAddNote_Click(ByVal sender As Object, ByVal e As EventArgs)
			Properties.AddNote(DateTime)
		End Sub

		Protected Overridable Sub OnMenuEditNote_Click(ByVal sender As Object, ByVal e As EventArgs)
			Properties.EditNotes(DateTime)
		End Sub

		Protected Overridable Sub OnMenuDeleteNote_Click(ByVal sender As Object, ByVal e As EventArgs)
			Properties.DeleteNotes(DateTime)
		End Sub
	End Class
End Namespace
